sys/fw_slots/fw_slots.c

```diff
@@ -0,0 +1,477 @@
+            fw_slots_print_metadata(&fw_slot_metadata);
+        } else {
+            printf("[fw_slots] ERROR cannot get slot metadata.\n");
+        }
+
+        /* Is this slot invalid? If yes, let's treat it as empty. */
+        if (fw_slots_validate_metadata(&fw_slot_metadata) == false) {
+            return slot;
+        }
+    }
+
+    printf("[fw_slots] Could not find any empty FW slots!"
+            "\nSearching for oldest FW slot...\n");
+    /*
+     * If execution goes this far, no empty slot was found. Now, we look for
+     * the oldest FW slot instead.
+     */
+    return fw_slots_find_oldest_int_image();
+}
+
+int fw_slots_find_oldest_int_image(void)
+{
+    /* The oldest firmware should be the v0 */
+    int oldest_fw_slot = 1;
+    uint16_t oldest_firmware_version = 0;
+
+    /* Iterate through each of the MAX_FW_SLOTS internal slots. */
+    for (int slot = 1; slot <= MAX_FW_SLOTS; slot++) {
+        /* Get the metadata of the current FW slot. */
+        FW_metadata_t fw_slot_metadata;
+
+        if(fw_slots_get_int_slot_metadata(slot, &fw_slot_metadata) == 0) {
+            fw_slots_print_metadata(&fw_slot_metadata);
+        } else {
+            printf("[fw_slots] ERROR cannot get slot metadata.\n");
+        }
+
+        /* Is this slot populated? If not, skip. */
+        if (fw_slots_validate_metadata(&fw_slot_metadata) == false) {
+            continue;
+        }
+
+        /* Is this the oldest image we've found thus far? */
+        if (oldest_firmware_version) {
+            if (fw_slot_metadata.version < oldest_firmware_version) {
+                oldest_fw_slot = slot;
+                oldest_firmware_version = fw_slot_metadata.version;
+            }
+        } else {
+            oldest_fw_slot = slot;
+            oldest_firmware_version = fw_slot_metadata.version;
+        }
+    }
+
+    printf("[fw_slots] Oldest FW slot: #%d; Firmware v%d\n", oldest_fw_slot,
+            oldest_firmware_version);
+
+    return oldest_fw_slot;
+}
+
+int fw_slots_find_newest_int_image(void)
+{
+    /* At first, we only assume knowledge of version v0 */
+    int newest_fw_slot = 0;
+    uint16_t newest_firmware_version = 0;
+
+    /* Iterate through each of the MAX_FW_SLOTS. */
+    for (int slot = 1; slot <= MAX_FW_SLOTS ; slot++) {
+        /* Get the metadata of the current FW slot. */
+        FW_metadata_t fw_slot_metadata;
+
+        if(fw_slots_get_int_slot_metadata(slot, &fw_slot_metadata) == 0) {
+            fw_slots_print_metadata(&fw_slot_metadata);
+        } else {
+            printf("[fw_slots] ERROR cannot get slot metadata.\n");
+        }
+
+        /* Is this slot populated? If not, skip. */
+        if (fw_slots_validate_metadata(&fw_slot_metadata) == -1) {
+            continue;
+        }
+
+        /* Is this the newest non-Golden Image image we've found thus far? */
+        if ( fw_slot_metadata.version > newest_firmware_version ) {
```

Comment:
This does not work, if a firmware with version number 0 is the latest version, e.g. if only one image is installed, because slot number 0 is an invalid slot. At least for the code used in the main.c file of the bootloader.
The easiest solution would be to define 0 as an invalid firmware version number in the documentation. But currently, you're using version number 0 in the Makefile of the firmware_swapping example.